lib/linked_list.rb

```diff
@@ -18,129 +19,255 @@ def initialize
+          return node if node.value == value
+          while (node = node.next)
+            return node if node.value == value
+          end
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n), where n is the length of the list
+    # Space Complexity: O(1)
     def find_max
-      raise NotImplementedError
+        if @head == nil
+              return nil
+             end
+             current = @head
+             max = current.data
+             while current
+              if current.data > max
+                max = current.data
+              end
+              current = current.next
+              end
+              return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n), where n is the length of the list
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+        if @head == nil
+                return nil
+              end
+              current = @head
+              min = current.data
+              while current
+               if current.data < min
+                 min = current.data
+               end
+               current = current.next
+               end
+               return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n), where n is the length of the list
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      current = @head
+
+      return 0 if current.nil?
+      length = 0
+
+      until current.nil?
+        current = current.next
+        length += 1
+      end
+
+      return length
     end
 
+    # method that inserts a given value as a new last node in the linked list
+        # Time Complexity: O(n), where n is the length of the list
+        # Space Complexity: unsure. O(1) because we're not creating a new data structure, we're just adding to
+        # an existing one?
+        def add_last(value)
+            new_node = Node.new(value)
+          if @head
+               current = @head
+                until current.next.nil?
+                  current = current.next
+                end
+              current.next = new_node
+          else
+              @head = new_node
+          end
+        end
+
+        # method that returns the value of the last node in the linked list
+        # returns nil if the linked list is empty
+        # Time Complexity: O(n), where n is the length of the list
+        # Space Complexity: O(1)
+        def get_last
+          node = @head
+
+          return nil if node.nil?
+
+          return node.data if !node.next
+          return node.data if !node.next while (node = node.next)

```

Comment:
Take a look at this line.  I strongly suggest you revise it so it's more readable.